git/obj_iter.go

```diff
@@ -2,128 +2,147 @@ package git
 
 import (
 	"bufio"
+	"context"
 	"fmt"
 	"io"
-	"os"
-	"os/exec"
+
+	"github.com/github/git-sizer/internal/pipe"
 )
 
 // ObjectIter iterates over objects in a Git repository.
 type ObjectIter struct {
-	cmd1    *exec.Cmd
-	cmd2    *exec.Cmd
-	out1    io.ReadCloser
-	out2    io.ReadCloser
-	f       *bufio.Reader
-	errChan <-chan error
+	ctx      context.Context
+	p        *pipe.Pipeline
+	oidCh    chan OID
+	errCh    chan error
+	headerCh chan BatchHeader
 }
 
 // NewObjectIter returns an iterator that iterates over objects in
 // `repo`. The arguments are passed to `git rev-list --objects`. The
 // second return value is the stdin of the `rev-list` command. The
 // caller can feed values into it but must close it in any case.
-func (repo *Repository) NewObjectIter(
-	args ...string,
-) (*ObjectIter, io.WriteCloser, error) {
-	cmd1 := repo.GitCommand(append([]string{"rev-list", "--objects"}, args...)...)
-	in1, err := cmd1.StdinPipe()
-	if err != nil {
-		return nil, nil, err
-	}
-
-	out1, err := cmd1.StdoutPipe()
-	if err != nil {
-		return nil, nil, err
-	}
-
-	cmd1.Stderr = os.Stderr
-
-	err = cmd1.Start()
-	if err != nil {
-		return nil, nil, err
-	}
-
-	cmd2 := repo.GitCommand("cat-file", "--batch-check", "--buffer")
-	in2, err := cmd2.StdinPipe()
-	if err != nil {
-		out1.Close()
-		cmd1.Wait()
-		return nil, nil, err
-	}
-
-	out2, err := cmd2.StdoutPipe()
-	if err != nil {
-		in2.Close()
-		out1.Close()
-		cmd1.Wait()
-		return nil, nil, err
+func (repo *Repository) NewObjectIter(ctx context.Context, args ...string) (*ObjectIter, error) {
```

Comment:
The `args` object seems to be unused now, with `--stdin` and `--date-order` hardcoded in the rev list command.  I don't really see the need to pass `args` in here, should that just be dropped from the signature?

Reply:
Nice catch! Fixed.